runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleMonitoringInfoBuilder.java

```diff
@@ -86,46 +89,97 @@ public SimpleMonitoringInfoBuilder setUrn(String urn) {
-    this.setInt64TypeUrn();
+  public SimpleMonitoringInfoBuilder setDoubleSumValue(double value) {
+    ByteString.Output output = ByteString.newOutput();
+    try {
+      DOUBLE_CODER.encode(value, output);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    this.builder.setPayload(output.toByteString());
+    this.builder.setType(MonitoringInfoConstants.TypeUrns.SUM_DOUBLE_TYPE);
     return this;
   }
 
   /**
-   * Sets the IntDistributionData of the DistributionData in the MonitoringInfo, and the appropriate
-   * type URN.
+   * Encodes the value and sets the type to {@link
+   * MonitoringInfoConstants.TypeUrns#LATEST_INT64_TYPE}.
    */
-  public SimpleMonitoringInfoBuilder setInt64DistributionValue(DistributionData data) {
-    this.builder
-        .getMetricBuilder()
-        .getDistributionDataBuilder()
-        .getIntDistributionDataBuilder()
-        .setCount(data.count())
-        .setSum(data.sum())
-        .setMin(data.min())
-        .setMax(data.max());
-    this.setInt64DistributionTypeUrn();
+  public SimpleMonitoringInfoBuilder setInt64LatestValue(GaugeData data) {
+    checkArgument(GaugeData.empty() != data, "Cannot encode empty gauge data");
+    ByteString.Output output = ByteString.newOutput();
+    try {
+      VARINT_CODER.encode(data.timestamp().getMillis(), output);
+      VARINT_CODER.encode(data.value(), output);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    this.builder.setPayload(output.toByteString());
+    this.builder.setType(MonitoringInfoConstants.TypeUrns.LATEST_INT64_TYPE);
     return this;
   }
 
-  /** Sets the the appropriate type URN for int64 distribution tuples. */
-  public SimpleMonitoringInfoBuilder setInt64DistributionTypeUrn() {
-    this.builder.setType(MonitoringInfoConstants.TypeUrns.DISTRIBUTION_INT64);
+  /**
+   * Encodes the value and sets the type to {@link
+   * MonitoringInfoConstants.TypeUrns#DISTRIBUTION_INT64_TYPE}.
+   */
+  public SimpleMonitoringInfoBuilder setInt64DistributionValue(DistributionData data) {
+    ByteString.Output output = ByteString.newOutput();
+    try {
+      VARINT_CODER.encode(data.count(), output);
+      VARINT_CODER.encode(data.sum(), output);
+      VARINT_CODER.encode(data.min(), output);
+      VARINT_CODER.encode(data.max(), output);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    this.builder.setPayload(output.toByteString());
+    this.builder.setType(MonitoringInfoConstants.TypeUrns.DISTRIBUTION_INT64_TYPE);
     return this;
   }
 
-  /** Sets the the appropriate type URN for sum int64 counters. */
-  public SimpleMonitoringInfoBuilder setInt64TypeUrn() {
-    this.builder.setType(MonitoringInfoConstants.TypeUrns.SUM_INT64);
+  /**
+   * Encodes the value and sets the type to {@link
+   * MonitoringInfoConstants.TypeUrns#DISTRIBUTION_INT64_TYPE}.
+   */
+  public SimpleMonitoringInfoBuilder setDoubleDistributionValue(
```

Comment:
Looks like SimpleMonitoringInfoBuilder helped a lot with this refactor :)

Reply:
Yup